src/core/observable.ts

```diff
@@ -16,46 +64,115 @@ export interface IObservable extends IDepTreeNode {
 	 * the dependency is already established
 	 */
 	lastAccessedBy: number;
-	staleObservers: IDerivation[];
-	observers: SimpleSet<IDerivation>;
+
+	isPendingUnobservation: boolean; // for effective unobserving
+	isObserved: boolean; // for unobserving only once ber observation
+	// sets of observers grouped their state to only notify about changes.
+	observers: DerivationsSets;
+
 	onBecomeUnobserved();
 }
 
 export function addObserver(observable: IObservable, node: IDerivation) {
 	observable.observers.add(node);
+	if (!observable.isObserved) {
+		observable.isObserved = true;
+	}
 }
 
 export function removeObserver(observable: IObservable, node: IDerivation) {
 	observable.observers.remove(node);
-	if (observable.observers.length === 0)
-		observable.onBecomeUnobserved(); // TODO: test if this happens only once, e.g. remove returns bool!
+	if (observable.isObserved && observable.observers.length === 0) {
+		if (globalState.inBatch > 0) {
+			/**
+			 * Wan't to observe/unobserve max once per observable during batch.
+			 * Let's postpone it to end of the batch
+			 */
+			if (!observable.isPendingUnobservation) {
+				observable.isPendingUnobservation = true;
+				globalState.pendingUnobservations.push(observable);
+			}
+		} else {
+			observable.isObserved = false;
+			observable.onBecomeUnobserved(); // TODO: test if this happens only once, e.g. remove returns bool!
+		}
+	}
+}
+
+export function startBatch() {
+	globalState.inBatch++;
+}
+
+export function endBatch() {
+	if (--globalState.inBatch === 0) {
+		globalState.pendingUnobservations.splice(0).forEach(observable => {
+			observable.isPendingUnobservation = false;
+			if (observable.isObserved && observable.observers.length === 0) {
+				observable.isObserved = false;
+				observable.onBecomeUnobserved(); // TODO: test if this happens only once, e.g. remove returns bool!
+			}
+		});
+	}
 }
 
 export function reportObserved(observable: IObservable) {
-	if (globalState.isTracking === false)
+	if (globalState.isTracking === false) {
+		if (globalState.inBatch > 0 && !observable.isObserved) {
+			// pseudoobserving by batch
+			observable.isObserved = true;
+			// probably will want to unobserve it at the end of the batch
+			if (!observable.isPendingUnobservation) {
+				observable.isPendingUnobservation = true;
+				globalState.pendingUnobservations.push(observable);
+			}
+		}
 		return;
+	}
 	const derivation = globalState.derivationStack[globalState.derivationStack.length - 1];
 	/**
 	 * Simple optimization, give each derivation run an unique id (runId)
 	 * Check if last time this observable was accessed the same runId is used
 	 * if this is the case, the relation is already known
 	 */
+	observable.isObserved = true;
```

Comment:
if i'm not mistaken, this can move inside the if. btw if that is the case, `addObserver` might not need to set `isObserved` as well?